ServerOptions:asOptionsString: fix to use correct concatenation

LFSaw opened the pull request:
This PR fixes `asOptionsString` to use correct concatenation (as described in `scsynth -h`) for `ugenPluginPath` (`scsynth -U <path>[:<path>[:<path>]]`) (`":"` instead of `"; "`).


## Test

Like most developers, I have built SC and sc3-plugins from source. This means I have two valid paths.


Test that pluginpath works

```
(
// plugin paths without plugins
s.options.ugenPluginsPath = thisProcess.platform.helpDir

)

s.options.asOptionsString;

s.reboot
```

results in (anticipated) errors

```
-> Booting server 'localhost' on address 127.0.0.1:57110.
*** ERROR: open directory failed '/localvol/sound/src/supercollider/Help': No such file or directory
exception in GraphDef_Load: UGen 'Control' not installed.
[...]
```

Test general plugin paths
```
(
// fill your own plugin paths here
s.options.ugenPluginsPath = [
	"/localvol/sound/src/sc3-plugins/build/build_osx/SC3plugins/",
	"/localvol/sound/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/plugins"
];
)

// escaped options string
s.options.asOptionsString;

s.reboot;
```

You should see something like this:

```
[...]
Found 98 LADSPA plugins
[...]
SuperCollider 3 server ready.
Requested notification messages from server 'localhost'
localhost: server process's maxLogins (1) matches with my options.
localhost: keeping clientID (0) as confirmed by server process.
Shared memory server interface initialized
```

This means sc3-plugins are found.

There should _not_ be anything like

```
exception in GraphDef_Load: UGen 'Control' not installed.
while reading file: '/home/dlm/.local/share/SuperCollider/synthdefs/system_line_audio.scsyndef'
```

This'd mean that the standard plugin library is _not_ loaded.

Alternatively, and particularly to test path escaping, make a hardcopy of the plugins to a temp directory that has spaces in its path and run

```
(
s.options.ugenPluginsPath = [
	"/Users/tboverma/tmp/pluginpath with space/SC3plugins/",
	"/Users/tboverma/tmp/pluginpath with space/plugins/"
];
)
```

Results should be similar as in the previous case.

brianlheim commented:
For Supernova, it may also be worth noting that the `-U` option may be specified several times, when you edit the options string.

LFSaw commented:
May I ask why windows has a different behavior? or why we rely on delimiter chars (`:`, `;`) that are incompatible between platforms?

brianlheim commented:
> May I ask why windows has a different behavior? or why we rely on delimiter chars (:, ;) that are incompatible between platforms?

This is the separator used for the PATH environment variable, so it's somewhat natural to use it for other lists of paths.

brianlheim commented:
See e.g. https://stackoverflow.com/questions/1499019/how-to-get-the-path-environment-variable-separator-in-python

telephon commented:
yes, and you could also add to `Platform`

```
UnixPlatform : Platform {
	pathSeparator { ^$/ }
        pathDelimiter { ^$: }
```

LFSaw commented:
thanks for the explanations and hints on implementation. I added the suggestions to the PR.

jamshark70 commented:
Should the `path:path:path` list be quoted or escaped? What if there is a space in one of the path names?

```
scsynth -U C:\Program Files\SuperCollider\plugins
```

^^ Will that work? I think we need quotes.

Side note:

At risk of sounding like a broken record, but while we are working to improve development processes lately, we should also consider that here, the original addition of `-U` to ServerOptions used `; ` (with a space!) as the delimiter, without quoting the path list (so the delimiter's space would break the list into separate C arguments). That is, this change must have gone in either without any testing in OSX or Linux, or testing only a single path. (With the space, I'm not even clear that multiple paths would have worked in Windows either.)

"Water under the bridge" but I've seen an awful lot of these over the years. Review could be greatly simplified if we could get in the habit of testing code not just for the simplest case.

LFSaw commented:
I don't have a Windows system so I cannot test this properly:
There are several possibilities to quote/escape the string, which one works on windows?

```supercollider
"/a/path/with spaces".shellQuote
"/a/path/with spaces".quote
"/a/path/with spaces".escapeChar($ )
```
For the linux/unix/osx implementation, I'd go for 

```supercollider
ugenPluginsPath.collect(_.escapeChar($ )).join(Platform.pathDelimiter)
```

However, if the windows way of doing it is different, this would need to be developed as a Platform-aware path-joiner, probably in `Platform`...

LFSaw commented:
@jamshark70 I'd be delighted to have a testing routine for this but am unfortunately not able to dive too deep into this atm... If you or someone else could make a suggestion where to look for the description of the current testing policy, I'm happy to take a glance. 

On second thought, the intended behaviour looks quite difficult to test to me (external process starting, no server snooping for actually loaded paths, etc.). Anyone has an idea on how to do that?

jamshark70 commented:
I can check Windows next time I'm in my office.

jamshark70 commented on SCClassLibrary/Common/Control/Server.sc:
```diff
@@ -141,7 +141,7 @@ ServerOptions {
 				ugenPluginsPath
 			} {
-				ugenPluginsPath.join("; ");
+				ugenPluginsPath.join(Platform.pathDelimiter);
```
User-specified paths might include spaces or other special characters. These should be escaped, or the paths quoted. IIRC Windows requires quoting.

```
-U my\ custom\ path:other\ path

or

-U 'my custom path':'other path'

or this? I don't know if the above or below is correct

-U 'my custom path:other path'
```

For that matter, I don't know if the paths should be quoted individually, or if the entire -U string should be quoted outside (with `:` or `;` inside the quotes).

LFSaw replied:
this is addressed

miguel-negrao commented:
> For the linux/unix/osx implementation, I'd go for
> 
> ugenPluginsPath.collect(_.escapeChar($ )).join(Platform.pathDelimiter)

In case the quoting is different in windows and unixes, then for unixes  the proper fix IMHO is to stop using String#unixCmd and switch to SequenceableCollection#unixCmd which doesn't need any quoting:

https://github.com/tai-studio/supercollider/blob/5df905b7794154c3d9e24225936ee76877f66eb0/SCClassLibrary/Common/Control/Server.sc#L924

	pid = unixCmd(program ++ options.asOptionsString(addr.port), { |exitCode|

If the quoting is the same then I guess it's less work to just quote in both cases...

gusano commented:
> Personally after seeing how much trouble this is causing I think it would be best to deprecate the multi-path form of this option and offload the parsing job onto boost::program_options the way supernova does

:+1:

LFSaw commented:
I'm all up for multiple -U option, How do we deal with it in the meantime? Atm, it is clearly broken (without this patch...)

gusano commented:
@LFSaw How many years has it been broken? :)

We could either:

1. open an issue to later implement the multiple `-U` options and merge your PR meanwhile
2. implement it directly leaving this broken for a bit more time

I'd vote for 1. as you already worked on the fix.

LFSaw commented:
> How many years has it been broken? :)

:) True, but now I stumbled over it because I need it :) 


> 1. open an issue to later implement the multiple -U options and merge your PR meanwhile

I'd love that (see above).

I implemented the quote addition that blheim suggested but need to do some testing (documentation). Will update the PR asap.

brianlheim commented:
That also sounds good to me. #3117 is the parent issue, I will leave a note there about this.

LFSaw commented:
I implemented the desired behaviour.

For testing, I followed HJH's advice from the sc-dev list and added the testing "routine" to the description of this PR (up).

LFSaw commented:
I tested on OSX and it works. 
I'd appreciate if someone with a windows machine could test and report back.

LFSaw commented:
@jamshark70 , you are the only one I know who has a windows machine, would you mind?

LFSaw commented:
@brianlheim I addressed your requests, could you re-evaluate, please?

telephon commented:
@brianlheim this should be OK as far as I can tell?

telephon commented:
@jamshark70 could you please check if your requests are fulfilled? 
Also Brian, take another quick look, please.

I think we (including me) should get into the habit of responding quickly to requests that were fulfilled, otherwise we will all stop to even respond, and eventually to even propose a solution.

jamshark70 commented:
Unfortunately, I can't test at the moment -- I had been using my office computer for Windows testing, but last week, a student broke the OS installation. So I don't have access to Windows now.

It looks OK to me.

> I think we (including me) should get into the habit of responding quickly to requests that were fulfilled, otherwise we will all stop to try and fulfill requests, and eventually to even propose a solution.

FWIW: I may have to start to discipline myself not to participate in development activities during the semester. It's simply gotten too busy. I can't keep up.

telephon commented:
>FWIW: I may have to start to discipline myself not to participate in development activities during the semester. It's simply gotten too busy. I can't keep up.

I completely understand. It seems also to be a problem of the github interface that it is difficult to track such things (I suppose one needs to configure when a notification is shown).

It might be good to share semester beginning and end between countries...
